flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotAssignerTest.java

```diff
@@ -0,0 +1,200 @@
+import static org.apache.flink.runtime.scheduler.adaptive.allocator.JobInformation.VertexInformation;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for {@link SlotAssigner}. */
+@ExtendWith(ParameterizedTestExtension.class)
+class SlotAssignerTest {
+
+    private static final TaskManagerLocation tml1 = new LocalTaskManagerLocation();
+    private static final SlotInfo slot1OfTml1 = new TestingSlot(tml1);
+    private static final SlotInfo slot2OfTml1 = new TestingSlot(tml1);
+    private static final SlotInfo slot3OfTml1 = new TestingSlot(tml1);
+
+    private static final TaskManagerLocation tml2 = new LocalTaskManagerLocation();
+    private static final SlotInfo slot1OfTml2 = new TestingSlot(tml2);
+    private static final SlotInfo slot2OfTml2 = new TestingSlot(tml2);
+    private static final SlotInfo slot3OfTml2 = new TestingSlot(tml2);
+
+    private static final TaskManagerLocation tml3 = new LocalTaskManagerLocation();
+    private static final SlotInfo slot1OfTml3 = new TestingSlot(tml3);
+    private static final SlotInfo slot2OfTml3 = new TestingSlot(tml3);
+
+    private static final List<SlotInfo> allSlots =
+            Arrays.asList(
+                    slot1OfTml1,
+                    slot2OfTml1,
+                    slot3OfTml1,
+                    slot1OfTml2,
+                    slot2OfTml2,
+                    slot3OfTml2,
+                    slot1OfTml3,
+                    slot2OfTml3);
+
+    private static final JobVertex jobVertex = new JobVertex("testingJobVertex");
+    private static final SlotSharingGroup slotSharingGroup = new SlotSharingGroup();
+
+    @Parameter SlotAssigner slotAssigner;
+
+    @Parameter(value = 1)
+    int parallelism;
+
+    @Parameter(value = 2)
+    Collection<? extends SlotInfo> freeSlots;
+
+    @Parameter(value = 3)
+    JobAllocationsInformation previousAllocations;
+
+    @Parameter(value = 4)
+    List<TaskManagerLocation> minimalTaskExecutors;
+
+    @TestTemplate
+    void testAssignSlots() {
+        final VertexInformation vertexInfo =
+                new TestVertexInformation(jobVertex.getID(), parallelism, slotSharingGroup);
+        final VertexParallelism vertexParallel =
+                new VertexParallelism(
+                        singletonMap(vertexInfo.getJobVertexID(), vertexInfo.getParallelism()));
+        final JobInformation jobInformation = new TestJobInformation(singletonList(vertexInfo));
+        final Set<TaskManagerLocation> keptTaskExecutors =
+                slotAssigner
+                        .assignSlots(jobInformation, freeSlots, vertexParallel, previousAllocations)
+                        .stream()
+                        .map(assignment -> assignment.getSlotInfo().getTaskManagerLocation())
+                        .collect(Collectors.toSet());
+        assertThat(minimalTaskExecutors).containsAll(keptTaskExecutors);
```

Comment:
As I understand, `containsAll` isn't expected here.

![](https://github.com/user-attachments/assets/7cf62b84-64a4-409d-8192-4fdab6094f44)

```suggestion
        assertThat(minimalTaskExecutors).containsExactlyElementsOf(keptTaskExecutors);
```

Reply:
Thanks @1996fanrui a lot for the comment.
updated ! :) ~